packages/strapi-admin/ee/admin/containers/Roles/CreatePage/Permissions/README.md

```diff
@@ -0,0 +1,377 @@
+
+## Concepts
+
+- Checkbox: has only 1 `checked=true` or `checked=false`.
+
+- Parent checkbox: A **parent checkbox** is a checkbox which the state value depends on the state of its children ones. It means that we cannot directly access its value from the `modifiedData` object.
+  Such checkbox can have multiple states:
+  - someChecked: `true` or `false`,
+  - checked: `true` of `false`
+    > Both states are coupled: if `someChecked=true` then `checked=false`.
+    > In terms of user's interaction when cliking a parent checkbox is will toggle the value of its children.
+
+Ex: given the following data:
+
+```js
+const modifiedData = {
+  address: {
+    create: {
+      fields: { f1: true, f2: true,}
+      locales: { en: false, fr: false}
+    },
+    update: {
+      enabled: true,
+    }
+  }
+}
+```
+
+From the `modifiedData` object we can identify 4 parent checkboxes:
+
+1. `address` which value depends on the values of `address.create` & `address.create.update` here the state will be: `someChecked=true`
+2. `create` which value depends on the values of `address.create.fields` & `address.create.locles` here the state will be: `someChecked=true, checked=true`
+3. `fields` which value depends on the values of `address.create.fields.f1` & `address.create.fields.f2` here the state will be: `checked=true`
+4. `locales` which value depends on the values of `address.create.locales.en` & `address.create.locales.fr` here the state will be `checked=false, someChecked=falses`
+
+> `address.update` is not a parent checkbox since we can access its value directly `address.update.enabled`
+
+---
+
+## Components architecture
+
+### `<Permissions />` archictecture
+
+```js
+<PermissionsDataManagerProvider>
+  <Tabs>
+    <ContentTypes /> // Used with the `layout.sections.collectionTypes` data
+    <ContentTypes /> // Used with the `layout.sections.singleTypes` data
+    <PluginsAndSettings /> // Used with the `layout.sections.settings` data
+    <PluginsAndSettings /> // Used with the `layout.sections.plugins` data
+  </Tabs>
+</PermissionsDataManagerProvider>
+```
+
+Below, is the architecture of the `<Permissions />` component:
+
+```js
+<ContentTypes>
+  <GlobalActions /> // Component in charge of displaying the global action checkboxes (parent
+  checkboxes), they are used to toggle all the checkboxes of the associated column
+  <ContentTypeCollapses>
+    {' '}
+    // Wrapper or the collapse and the matrix
+    <ContentTypeCollapse>
+      <Collapse /> // Main row of a content type => Displays the main actions of a content type
+      (parent checkboxes)
+      <CollapsePropertyMatrix>
+        {' '}
+        // Matrix of the actions subject property
+        <Header /> // Row that displays the actions labels inside a property (ex: create, read,
+        update)
+        <ActionRow>
+          {' '}
+          // Displays a subject's property values
+          <SubActionRow /> // Recursive component if a property has a children key, the component
```

Comment:
What is the difference between `ActionRow` and `SubActionRow`? Couldn't ActionRow be a SubActionRow?

Reply:
That was done to facilitate the recursive UI for nested components. Level 1 has a different UI than level 1 + n. Also, lvl 1+n doesn't have the left "global" checkbox.

For example, in the screenshot below, the red square is the ActionRow and the black square represent SubActionRow recursive

![Capture d’écran 2021-02-15 à 10 25 54](https://user-images.githubusercontent.com/7756284/107928331-75520400-6f78-11eb-8b6e-f6671597fb78.png)

Reply:
Not really because the SubActionRow displays a timeline.